src/snowflake/snowpark/_internal/utils.py

```diff
@@ -669,3 +669,64 @@ def get_copy_into_table_options(
         elif k not in NON_FORMAT_TYPE_OPTIONS:
             file_format_type_options[k] = v
     return file_format_type_options, copy_options
+
+
+def convert_ids_to_upper_case(name: str) -> str:
+    """convert a fully qualfied name to upper case when needed. The double-quoted part won't be converted.
+    This function doesn't validate whether the name has the right pattern.
+    Instead, it assumes the name has the right pattern.
+    """
+    i = 0
+
+    def skip_space() -> None:
+        nonlocal i
+        while i < len(name) and name[i] == " ":
+            i += 1
+
+    def find_next_id() -> str:
+        skip_space()
+        nonlocal i
+        if i >= len(name):
+            return None
+        if name[i] == '"':
+            quoted = True
+        else:
+            quoted = False
+        iden = name[i]
+        i += 1
+        if iden == ".":  # the 2nd dot in format a..b
+            return iden
+        while i < len(name):
+            if quoted:
+                if i + 1 < len(name) and name[i + 1] == '"' and name[i] == '"':
+                    iden += name[i : i + 2]
+                    i += 2
+                else:
+                    iden += name[i]
+                    if name[i] == '"':
+                        i += 1  # skip the close quote
+                        skip_space()
+                        i += 1  # skip the first dot after quote
+                        return iden
+                    else:
+                        i += 1
+            else:
+                if name[i] == ".":
+                    i += 1
+                    skip_space()
+                    return iden.upper().rstrip()
+                else:
+                    iden += name[i]
+                    i += 1
+        if iden:
+            return iden.strip().upper()
+        return iden
```

Comment:
I think we can just remove the if block and return `iden.strip().upper()` here.
Is this case only hit for the last unquoted identifier?

Reply:
+1